### PR TITLE
Refactoring of ComplexGraphType.Field* APIs

### PR DESCRIPTION
**Pull request by sungam3r**
A little revolution in `FieldXXX` APIs. A lot of methods are marked with `Obsolete` attribute. Only few methods left to allow construct field builder and then operate via its APIs. Splitted into separate PRs:
- #3242 
- #3244
- #3245 
- #3275

The goal is to simplify APIs and guide developer with well-discovered APIs. In some regard helps to solve #3046.

One of my ideas is that `Field()` should **declare** a field with minimal efforts, i.e. there should be no `FieldAsync`/`FieldDelegate`/`FieldSubscribe`/whatever. The intent of those methods is to set custom resolver. We already have a bunch of methods to specify different reolvers so I switched to them.

Also note that proposed approach looks more elegant and requires less lines of code.

\* Most of the changes are tests.

### CODE REVIEW

**Comment by sungam3r** [comment]
@Shane32 I think this is the last bits I would like to include in v7.

**Comment by sungam3r** [comment]
I'll add documentation update tomorrow.

**Comment by Shane32** [comment]
I need time to review. In general I’d say this is a tough pill to swallow. Not because I disagree with the API changes, but more because GraphQL.NET has for so very long used the old Field syntax. To deprecate the old syntax, saying it will be completely gone in v8 seems questionable to say the least. I’ll give more thoughts tomorrow after I have time to actually review all the changes here and consider the implications.

**Comment by sungam3r** [comment]
If one wants to stay on old APIs then he can copy-paste old methods since all of them are just wrappers around "base" APIs. For a very long time I did not like all that bunch of methods that seemed to be needed for the same thing, but in which I constantly got confused. Somewhere you need to specify nullas an unused parameter (description arg somewhere), somewhere you should decide whether use generic argument or not. Field or FieldAsync / etc 😕 .

File: src/GraphQL/Types/Composite/ComplexGraphType.cs [hunk 1]
```diff
@@ -129,6 +102,14 @@ public virtual FieldType AddField(FieldType fieldType)
             return fieldType;
         }
 
+        /// <summary>
+        /// Creates a field builder used by all Field() methods.
+        /// </summary>
+        protected virtual FieldBuilder<TSourceType, TReturnType> CreateBuilder<TReturnType>(Type type)
```
Comment (sungam3r) [inline src/GraphQL/Types/Composite/ComplexGraphType.cs:108]: @Shane32 A new feature here to allow change default field builder in one place.

**Comment by Shane32** [comment]
I think we should split this PR up a bit to ease review process.

1. Enhance FieldBuilder methods
2. Rewrite all samples and documentation to use field builder methods
3. Deprecate old Field methods

**Comment by sungam3r** [comment]
1 won't compile without 2.

**Comment by Shane32** [comment]
You mean 2 won't compile without 1 ?

**Comment by sungam3r** [comment]
Yes. I deleted comment. Well, I would not to split, it will require additional time for me to jump around all changes again sorting them to different PRs. There are a lot of changes but almost all of them are 1-liners and rather simple to track:

изображение

**Comment by sungam3r** [comment]
I would say that there was more mechanical work on rewriting existing code (tests) than the work of the brain to write a new code 😄 .

**Comment by Shane32** [comment]
Let's at least move the `[Obsolete]` tags into a separate PR.

File: src/GraphQL.ApiTests/net6/GraphQL.approved.txt [hunk 2]
```diff
@@ -1751,24 +1765,52 @@ namespace GraphQL.Types
...
+        protected virtual GraphQL.Builders.FieldBuilder<TSourceType, TReturnType> CreateBuilder<TReturnType>(System.Type type) { }
+        public virtual GraphQL.Builders.FieldBuilder<TSourceType, object> Field(string name, GraphQL.Types.IGraphType type) { }
+        public virtual GraphQL.Builders.FieldBuilder<TSourceType, object> Field(System.Type type, string name = "default") { }
+        [System.Obsolete("Please use one of the Field() methods returning FieldBuilder and then methods def" +
+            "ined on it or just use AddField() method directly. This method will be removed i" +
+            "n v8.")]
         public GraphQL.Types.FieldType Field(System.Type type, string name, string? description = null, GraphQL.Types.QueryArguments? arguments = null, System.Func<GraphQL.IResolveFieldContext<TSourceType>, object?>? resolve = null, string? deprecationReason = null) { }
-        public virtual GraphQL.Builders.FieldBuilder<TSourceType, object> Field<TGraphType>()
+        public virtual GraphQL.Builders.FieldBuilder<TSourceType, object> Field<TGraphType>(string name = "default")
```
[the remark this answers is not in the document]
Comment (sungam3r) [reply src/GraphQL.ApiTests/net6/GraphQL.approved.txt:1776]: Yes, a "little" breaking change. I added note.

**Comment by sungam3r** [comment]
Changed files became a little less.

**Comment by Shane32** [comment]
Looking good.  Let's move all the `[Obsolete]` tags and changes to the migration document to a new PR.  Then we can get all the rest of documentation changes and changes to tests merged in beforehand.

**Review by Shane32 — COMMENTED** [review]
I looked through almost all files

File: docs2/site/docs/getting-started/dependency-injection.md [hunk 3]
```diff
@@ -298,15 +296,13 @@ public class StarWarsQuery : ObjectGraphType
...
-            resolve: context =>
+        Field<DroidType>("hero")
+            .Resolve(context =>
             {
                 using var scope = context.RequestServices.CreateScope();
                 var services = scope.ServiceProvider;
                 return services.GetRequiredService<MyDbContext>().Droids.Find(1);
-            }
-        );
+            });
```
[the remark this answers is not in the document]
Comment (sungam3r) [reply docs2/site/docs/getting-started/dependency-injection.md:305]: Do you want to post PR after?

[the remark this answers is not in the document]
Comment (Shane32) [reply docs2/site/docs/getting-started/dependency-injection.md:305]: Fine with me.

File: docs2/site/docs/getting-started/schema-types.md [hunk 4]
```diff
@@ -206,9 +206,9 @@ public class EpisodeEnum : EnumerationGraphType
     {
         Name = "Episode";
         Description = "One of the films in the Star Wars Trilogy.";
-        AddValue("NEWHOPE", "Released in 1977.", 4);
-        AddValue("EMPIRE", "Released in 1980.", 5);
-        AddValue("JEDI", "Released in 1983.", 6);
+        AddValue("NEWHOPE", 4, "Released in 1977.");
+        AddValue("EMPIRE", 5, "Released in 1980.");
+        AddValue("JEDI", 6, "Released in 1983.");
```
[the remark this answers is not in the document]
Comment (sungam3r) [reply docs2/site/docs/getting-started/schema-types.md:211]: fixed in part 2 PR

File: src/GraphQL.ApiTests/net6/GraphQL.approved.txt [hunk 5]
```diff
@@ -825,30 +825,44 @@ namespace GraphQL.Builders
...
+    [System.Obsolete("This class will be removed in v8.")]
     public static class FieldBuilder
     {
+        [System.Obsolete("Please use FieldBuilder<TSourceType, TReturnType>.Create() method. This method wi" +
+            "ll be removed in v8.")]
         public static GraphQL.Builders.FieldBuilder<TSourceType, TReturnType> Create<TSourceType, TReturnType>(GraphQL.Types.IGraphType type) { }
+        [System.Obsolete("Please use FieldBuilder<TSourceType, TReturnType>.Create() method. This method wi" +
+            "ll be removed in v8.")]
         public static GraphQL.Builders.FieldBuilder<TSourceType, TReturnType> Create<TSourceType, TReturnType>(System.Type? type = null) { }
     }
```
[the remark this answers is not in the document]
Comment (sungam3r) [reply src/GraphQL.ApiTests/net6/GraphQL.approved.txt:837]: Only obsolete annotation stay here after merging part 2 PR.

File: src/GraphQL/Builders/FieldBuilder.cs [hunk 6]
```diff
@@ -58,7 +58,6 @@ protected FieldBuilder(FieldType fieldType)
             {
                 Name = name,
                 ResolvedType = type,
-                Arguments = new QueryArguments(),
```
Comment (sungam3r) [inline src/GraphQL/Builders/FieldBuilder.cs:61]: @Shane32 What do you think about thee 2 lines left?

Comment (Shane32) [reply src/GraphQL/Builders/FieldBuilder.cs:61]: No comment.  Seems fine 🤷‍♂️

File: docs2/site/docs/getting-started/dependency-injection.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Shane32) [reply docs2/site/docs/getting-started/dependency-injection.md]: #3274

[the remark this answers is not in the document]
Comment (sungam3r) [reply docs2/site/docs/getting-started/dependency-injection.md]: I see a lot of stuff to deal with in my mailbox. I propose to review/merge as much as possible befor v7. I'll review tomorrow.

[the remark this answers is not in the document]
Comment (Shane32) [reply docs2/site/docs/getting-started/dependency-injection.md]: I agree and will flag the relevant open PRs with the 7.0 milestone.

File: src/GraphQL/Builders/FieldBuilder.cs (hunk 6 above)
Comment (sungam3r) [reply src/GraphQL/Builders/FieldBuilder.cs:61]: #3275

**Comment by sungam3r** [comment]
Nothing to push here. Closing.

**Pull request closed without merging**